src/main/java/mate/academy/security/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,42 @@
+package mate.academy.security.service.impl;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.security.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDbOptional = userService.findByEmail(email);
+        if (userFromDbOptional.isPresent()) {
+            User userFromDb = userFromDbOptional.get();
+            if (password != null) {
+                String hashedPassword = HashUtil.hashPassword(password, userFromDb.getSalt());
+                if (userFromDb.getPassword().equals(hashedPassword)) {
+                    return userFromDb;
+
+                }
+            }
+        }
+        throw new AuthenticationException("Email or password is incorrect");
+    }
```

Comment:
I think it's too complicated
```suggestion
        Optional<User> userFromDbOptional = userService.findByEmail(email);
        User userFromDb = userFromDbOptional.orElseThrow(() ->  new AuthenticationException("Email or password is incorrect"));
        if (password != null) { // I think password can't be null in DB, what do u think?
            String hashedPassword = HashUtil.hashPassword(password, userFromDb.getSalt());
            if (userFromDb.getPassword().equals(hashedPassword)) {
                return userFromDb;
            }
```

Reply:
If optional is not empty, then when I check if the hashes are equals, I will again need to throw exception if they are not equal, because otherwise the code will not compile (because all "if" constructions may not be executed). In my code I check all the conditions  step by step and throw an exception if something doesn't true. And if the password is null, then the method "hashPassword" will throw NPE, so I need to check it